crates/fuel-core/src/graphql_api/api_service.rs

```diff
@@ -346,11 +361,73 @@ async fn health() -> Json<serde_json::Value> {
     Json(json!({ "up": true }))
 }
 
+#[derive(Clone)]
+struct RequiredHeight(Option<u32>);
+
+#[async_trait::async_trait]
+impl<Body> FromRequest<Body> for RequiredHeight
+where
+    Body: Send,
+{
+    type Rejection = (StatusCode, String);
+
+    async fn from_request(
+        req: &mut axum::extract::RequestParts<Body>,
+    ) -> Result<Self, Self::Rejection> {
+        let required_fuel_block_height = req
+            .headers()
+            .get(REQUIRED_FUEL_BLOCK_HEIGHT_HEADER)
+            .map(|value| value.to_str())
+            .transpose()
+            .map_err(|_| (StatusCode::BAD_REQUEST, "Header Malformed".to_string()))?
+            .map(|value| value.parse::<u32>())
+            .transpose()
+            .map_err(|_| (StatusCode::BAD_REQUEST, "Header Malformed".to_string()))?;
+
+        Ok(RequiredHeight(required_fuel_block_height))
+    }
+}
+
 async fn graphql_handler(
+    extract_height: RequiredHeight,
     schema: Extension<CoreSchema>,
     req: Json<Request>,
-) -> Json<Response> {
-    schema.execute(req.0).await.into()
+) -> Result<Json<Response>, (StatusCode, http::HeaderMap, Json<String>)> {
+    let mut request = req.0;
+
+    if let RequiredHeight(Some(height)) = extract_height {
+        request
+            .extensions
+            .insert(REQUIRED_FUEL_BLOCK_HEIGHT_HEADER.to_string(), height.into());
+    };
+
+    let current_fuel_block_height_data: Arc<Mutex<Option<BlockHeight>>> =
+        Arc::new(Mutex::new(None));
+
+    let request = request.data(current_fuel_block_height_data.clone());
+    let graphql_response: Response = schema.execute(request).await.into();
+    let precondition_failed_error = graphql_response.errors.first().filter(|err| {
+        err.source::<RequiredFuelBlockHeightTooFarInTheFuture>()
+            .is_some()
+    });
+    if precondition_failed_error.is_some() {
+        let mut header_map = http::HeaderMap::new();
+        {
+            let current_fuel_block_height = current_fuel_block_height_data.lock().await;
+
+            header_map.insert(
+                CURRENT_FUEL_BLOCK_HEIGHT_HEADER,
+                (*current_fuel_block_height.expect("Block height is set")).into(),
+            );
+        }
+        Err((
+            StatusCode::PRECONDITION_FAILED,
+            header_map,
```

Comment:
todo: need to test that this does not get wiped out when processing the response upstream